File: Sliding Window/0340. Longest Substring with At Most K Distinct Characters/0340.py
class Solution:
    def lengthOfLongestSubstringKDistinct(self, s: str, k: int) -> int:
        freq = defaultdict(int)
        l=res=distinct=0

        for r in range(len(s)):
            freq[s[r]]+=1
            if freq[s[r]]==1:
                distinct+=1
            while distinct>k:
                freq[s[l]]-=1
                if freq[s[l]]==0:
                    distinct-=1
                l+=1
            if distinct<=k:
                res=max(res, r-l+1)
        return res
            

File: Sliding Window/0340. Longest Substring with At Most K Distinct Characters/0340.cpp
class Solution {
public:
    int lengthOfLongestSubstringKDistinct(string s, int k) {
        ios_base::sync_with_stdio(0);
        cin.tie(0);
        cout.tie(0);


        unordered_map<char,int> freq;
        int res=0;
        int distinct=0;
        int l=0;
        
        for(int r=0; r<s.size(); ++r){
            freq[s[r]]++;
            if (freq[s[r]]==1){
                distinct++;
            }
            while(distinct>k){
                freq[s[l]]--;
                if (freq[s[l]]==0){
                    distinct--;
                }
                l++;
            }
            if(distinct<=k){
                res=max(res,r-l+1);
            }
        }
        return res;

    }
};